samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,295 @@
+    }
+  }
+
+  static final class GraphNodes {
+    private List<Job> jobs;
+    private List<StreamNode> streams;
+
+    public List<Job> getJobs() {
+      return jobs;
+    }
+
+    public void setJobs(List<Job> jobs) {
+      this.jobs = jobs;
+    }
+
+    public List<StreamNode> getStreams() {
+      return streams;
+    }
+
+    public void setStreams(List<StreamNode> streams) {
+      this.streams = streams;
+    }
+  }
+
+  private int nextID = 0;
+  // streamId to StreamNode mapping
+  private final Map<String, StreamNode> streamNodes = new HashMap<>();
+  // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+  // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
+    StreamGraphImpl streamGraph = (StreamGraphImpl) jobGraph.getStreamGraph();
+    GraphNodes nodes = new GraphNodes();
+    List<Job> jobs = jobGraph.getJobNodes().stream()
+        .map(jobNode -> buildJob(jobNode, streamGraph, jobGraph))
+        .collect(Collectors.toList());
+    nodes.setJobs(jobs);
+    nodes.setStreams(createListAndSort(streamNodes.values()));
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    ObjectMapper mapper = new ObjectMapper();
+    mapper.writeValue(out, nodes);
+    return new String(out.toByteArray());
+  }
+
+  private Job buildJob(JobNode jobNode, StreamGraphImpl streamGraph, JobGraph jobGraph) {
+    Map<OutputStream, StreamEdge> outputToStreamSpecs = streamGraph.getOutStreams().entrySet().stream().collect(
+        Collectors.toMap(Map.Entry::getValue, e -> jobGraph.getOrCreateEdge(e.getKey())));
+    Map<Integer, OpNode> opNodes = new HashMap<>();
+
+    streamGraph.getInStreams().forEach((streamSpec, stream) -> {
+        StreamEdge streamEdge = jobGraph.getOrCreateEdge(streamSpec);
+        GraphNode node = getOrCreateStreamNode(streamEdge);
+        buildJsonGraph(stream, node, opNodes, outputToStreamSpecs);
+      });
+
+    Job job = new Job();
+    job.setJobName(jobNode.getJobName());
+    job.setJobId(jobNode.getJobId());
+    job.setOpNodes(createListAndSort(opNodes.values()));
+    return job;
+  }
+
+  private void buildJsonGraph(MessageStream inputStream, GraphNode node, Map<Integer, OpNode> opNodes, Map<OutputStream, StreamEdge> outputToStreamSpecs) {
+    Collection<OperatorSpec> specs = ((MessageStreamImpl) inputStream).getRegisteredOperatorSpecs();
+    for (OperatorSpec spec : specs) {
+      GraphNode next = getOrCreateOpNode(opNodes, spec);
+      node.getSuccessors().add(next.getNodeId());
+
+      if (spec instanceof SinkOperatorSpec) {
+        StreamEdge streamEdge = outputToStreamSpecs.get(((SinkOperatorSpec) spec).getOutStream());
```

Comment:
getOutStream() was renamed to getNextStream() in #92 to distinguish b/w actual/physical output streams and the next MessageStream in the operator DAG. The OutputStream in SinkOperatorSpec is the one that actual output is sent to. It has no next stream since it's a terminal operator. partitionBy returns a new InputStream to chain further operations on. Does that make sense?